oslc4j-jena-provider/src/main/java/org/eclipse/lyo/oslc4j/provider/jena/UniversalResourceSingleProvider.java

```diff
@@ -0,0 +1,92 @@
+/*!*****************************************************************************
+ * Copyright (c) 2019 Andrew Berezovskyi.
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * and Eclipse Distribution License v. 1.0 which accompanies this distribution.
+ *
+ * The Eclipse Public License is available at http://www.eclipse.org/legal/epl-v10.html
+ * and the Eclipse Distribution License is available at
+ * http://www.eclipse.org/org/documents/edl-v10.php.
+ *
+ * Contributors:
+ *
+ *	   Andrew Berezovskyi   - initial implementation
+ *******************************************************************************/
+package org.eclipse.lyo.oslc4j.provider.jena;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.lang.annotation.Annotation;
+import java.lang.reflect.Type;
+import javax.ws.rs.Consumes;
+import javax.ws.rs.Produces;
+import javax.ws.rs.WebApplicationException;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.MultivaluedMap;
+import javax.ws.rs.ext.MessageBodyReader;
+import javax.ws.rs.ext.MessageBodyWriter;
+import javax.ws.rs.ext.Provider;
+import org.eclipse.lyo.oslc4j.core.model.IResource;
+import org.eclipse.lyo.oslc4j.core.model.OslcMediaType;
+
+/**
+ * A new JAX-RS provider to (un)marshall IResource's from/to most common RDF formats.
+ * <p>
+ * JMH accepts objects in general but there is a slight conflict in the JAX pipeline with the
+ * Collections and Arrays, which are technically also Objects. If someone needs raw Objects, they
+ * can use old Providers.
+ *
+ * @since 4.0.0
+ */
+@Provider
+@Produces({OslcMediaType.APPLICATION_JSON_LD, OslcMediaType.TEXT_TURTLE,
+        OslcMediaType.N_TRIPLES_MIME, OslcMediaType.RDF_JSON_MIME, OslcMediaType.RDF_THRIFT_MIME})
+@Consumes({OslcMediaType.APPLICATION_JSON_LD, OslcMediaType.TEXT_TURTLE,
+        OslcMediaType.N_TRIPLES_MIME, OslcMediaType.RDF_JSON_MIME, OslcMediaType.RDF_THRIFT_MIME})
+public class UniversalResourceSingleProvider extends AbstractRdfProvider
+        implements MessageBodyReader<IResource>, MessageBodyWriter<IResource> {
```

Comment:
Can one also add?
implements MessageBodyReader<Collection<? extends IResource>>, MessageBodyWriter<Collection<? extends IResource>>

Reply:
No, unfortunately this does not work due to type erasure in Java. See https://softwareengineering.stackexchange.com/questions/219427/implementing-multiple-generic-interfaces-in-java for the background.

Reply:
Ok. We simply need to provide other Providers for collection.
Don’t even remember why we said we can do it all in 1 class.